Update YOLOv5 postprocess

furiosamg opened the pull request:
## Changes
- Switched to Ziglang for building manylinux wheels
- Renamed YOLOv5 to YOLO for version-agnostic postprocessing
- Updated Rust toolchain version
- Added and used `add_submodule` utility function
- Enhanced YOLO postprocess:
  - Defaults to class-aware NMS (can be turned off with `agnostic=True`)
  - Populates boxes when interpretable into multiple classes ([multi-label NMS](https://github.com/ultralytics/yolov5/blob/ace3e02e406307ed0cebc25aebe15835767107b2/utils/general.py#L356))
  - Update codes to avoid boundary checks
  - Returns `List[ndarray[shape=(num_dets, 6)]` instead of ODResult class list
  - Now stores agnostic parameter within postprocessor class
  - => Better accuracy ☺️

furiosamg committed:
Update import granularity

furiosamg committed:
Update toolchain & dependencies

furiosamg committed:
Copy argmax from rulinalg

furiosamg committed:
Do not reflect object conf score

furiosamg committed:
Remove unused Makefile commands

furiosamg committed:
Use zigbuild for manylinux build

furiosamg committed:
Remove unused Dockerfiles

furiosamg committed:
Stable sort

furiosamg committed:
Trim before performing nms & skip if low cls conf

furiosamg committed:
Bump version up to 0.10.0-dev0

furiosamg committed:
Sort in reverse order for box sorting

furiosamg committed:
Unstable sort

furiosamg committed:
Populate multiple class confidences

furiosamg committed:
Perform class-aware NMS

furiosamg committed:
Add module-related util function, __version__

furiosamg committed:
Use unchecked get function for speed

furiosamg committed:
Rename to yolo, unchecked access for speed

furiosamg committed:
Update submodule names

donghun-furiosa commented on src/yolo/mod.rs:
```diff
@@ -0,0 +1,305 @@
+        epsilon: Option<f32>,
+        agnostic: Option<bool>,
+    ) -> Vec<Array2<f32>> {
+        let max_nms: usize = 30_000;
```
```const```는 어떨까요

furiosamg replied:
감사합니다! :) c12e57a

furiosamg committed:
Make MAX_NMS as constant

donghun-furiosa commented on src/yolo/utils.rs:
```diff
@@ -17,7 +17,7 @@ pub struct DetectionBoxes {
     pub scores: Array1<f32>,
-    pub classes: Array1<usize>,
+    pub classes: Array1<f32>,
     pub len: usize,
```
이 pr 범위의 밖이기는 한데 len 가 있을필요가 없을것 같습니다. x1.len() = x2.len()  등등은 길이가 다 같은게 보장되는게 아닐까요?

furiosamg replied:
네 보장되는 것 맞습니다. 만든 이유는 밖에서 detection_boxes.(x1, x2, ...).len() 으로 꺼내 쓰는 게 안 이뻐서인데요.. 그냥 boxes.x1.len() 으로 바꾸는 것이 나을까요? ㅋㅋㅋㅋ

donghun-furiosa replied:
제 개인적인 생각은 DetectionBoxex 에 len 함수를 추가해서 boxes.len()을 호출하는것도 깔끔할것 같습니다.

furiosamg replied:
afb21d2ed0 에서 고쳤습니다.

donghun-furiosa commented on src/yolo/mod.rs:
```diff
@@ -0,0 +1,305 @@
+                }
+                results
+            })
+            .collect()
```
제가 추측(?)하기로는 par_iter 사용하시는건 rayon으로 multiprocessing 하시려는 의도인것 같은데 그렇다면 
```
  detection_boxes
            .into_par_iter()
            .map(|mut dbox| {
                if dbox.len > max_nms {
                    dbox.sort_by_score_and_trim(max_nms);
                };
                let indices =
                    Self::nms(&dbox, iou_threshold, epsilon, agnostic.unwrap_or(self.agnostic));
                let mut results = unsafe { Array2::uninit((indices.len(), 6)).assume_init() };
                for (i, &j) in indices.iter().enumerate() {
                    unsafe {
                        *results.uget_mut([i, 0]) = *dbox.x1.uget(j);
                        *results.uget_mut([i, 1]) = *dbox.y1.uget(j);
                        *results.uget_mut([i, 2]) = *dbox.x2.uget(j);
                        *results.uget_mut([i, 3]) = *dbox.y2.uget(j);
                        *results.uget_mut([i, 4]) = *dbox.scores.uget(j);
                        *results.uget_mut([i, 5]) = *dbox.classes.uget(j);
                    }
                }
                results
            })
            .collect()
```
방식으로 위에 indices 랑 묶어서 처리를 하면 어떨까요? + iter 대신 consuming 하는 into_iter 로 변경해보았습니다.

donghun-furiosa replied:
개인적으로 detectionbox -> Array2<f32> 로 가는것도 일종의 형변환만 존재해서(값 변동 x) 이 부분도 into 같은 함수로 빼도 깔끔해질것 같습니다.

furiosamg replied:
시그니쳐 상으로는 리뷰 주신 내용이 맞는데요,
지금 동작은 detectionbox + indexing list -> Array2<f32> 여서요
그러면 중간에 값을 한번 더 만들어야 할 것 같은데
detectionbox + indexing list -> (sub)detectionbox -> Array2
제가 제대로 이해한 게 맞을까요?

donghun-furiosa replied:
detectionbox에 to_array2(&self, iou_threshold, epsilon, agnostic.unwrap_or(self.agnostic)) 라는 함수를 추가하고 그 안에서 nms 에 계산도 하는걸 생각했습니다. (nms 자체가 RustPostProcessor 에 binding 될 필요도 없는것 같습니다). 다만 제가 말씀드린 방법보다 explicit 하게 형변환이 흐름상 더 좋다고 판단하신다면 into 같은 함수는 필요 없을것 같습니다.

furiosamg replied:
3c4a7bfd851 에서 고쳤습니다.

furiosamg committed:
Use maturin~=1.2.0 for resolving workaround

https://github.com/rust-cross/cargo-zigbuild/pull/140
https://github.com/PyO3/maturin/blob/v1.2.0/Cargo.toml#L84

furiosamg committed:
Implement len(), is_empty() for DetectionBoxes

furiosamg committed:
Parallelize NMS, chain processings

furiosamg committed:
Functionize select_and_convert

donghun-furiosa commented on src/yolo/mod.rs:
```diff
@@ -0,0 +1,305 @@
+                            // Find candidates where `class_confidence * object_confidence > conf_threshold`
+                            let candidates = class_confs
+                                .iter()
+                                .enumerate()
```
enumerate 일 필요가 없을것 같습니다!

furiosamg replied:
103번째 줄에서 꺼내 씁니다!

donghun-furiosa commented on src/yolo/mod.rs:
```diff
@@ -0,0 +1,305 @@
+                                scores.push(class_conf * object_confidence);
+                                classes.push(class_idx as f32);
+
+                                num_rows += 1;
+                                if num_rows >= MAX_BOXES {
+                                    break 'outer;
```
num_rows 가 0 시작이면 num_rows 가 max_boxes 일때 break 하니까 실제로는 max_boxes +1 까지 추가될것 같습니다

furiosamg replied:
a7999c176fd1 에서 고쳤습니다.

donghun-furiosa commented on src/yolo/utils.rs:
```diff
@@ -0,0 +1,106 @@
+    pub x2: Array1<f32>,
+    pub y2: Array1<f32>,
+    pub scores: Array1<f32>,
+    pub classes: Array1<f32>,
```
이건 그냥 궁금한 부분에 가까운데요. classes 가 int 가 아닌 f32인 이유가 있을까요? nms 계산할때 f32로 변환이 필요하기는 한데 처음 이 class 의 값을 usize -> f32로 변환이 일어나서 여쭤봅니다!

furiosamg replied:
원래는 usize Array 였었는데요, 말씀해주신대로 class-aware NMS 를 수행하도록 바꾸면서 f32 로 type cast가 어차피 일어나기 때문에 미리 f32로 바꿔 두었습니다. (나중에 전체 값이 f32 array 로 나가기 때문에 필요하기도 합니다)

furiosamg committed:
Limit boxes properly

donghun-furiosa reviewed (approved):
수고하셨습니다~

furiosamg commented:
@hyunsik 님, 1차 리뷰가 끝나서 머지를 위한 리뷰 부탁드리겠습니다.

hyunsik reviewed (approved):
제가 메시지를 놓쳐서 리뷰가 너무 늦었네요. 죄송합니다. 고생 많으셨습니다. 더 잘 정돈된 것 같습니다. 머지하셔도 좋을 것 같습니다.
